src/raftstore/store/worker/region.rs

```diff
@@ -250,22 +364,89 @@ impl SnapContext {
         timer.observe_duration();
     }
 
-    fn handle_destroy(&mut self, region_id: u64, start_key: Vec<u8>, end_key: Vec<u8>) {
-        info!(
-            "[region {}] deleting data in [{}, {})",
-            region_id,
-            escape(&start_key),
-            escape(&end_key)
-        );
+    fn cleanup_range(
+        &mut self,
+        region_id: u64,
+        start_key: Vec<u8>,
+        end_key: Vec<u8>,
+        use_delete_files: bool,
+    ) {
+        if use_delete_files {
+            if let Err(e) = util::delete_all_files_in_range(&self.kv_db, &start_key, &end_key) {
+                error!(
+                    "[region {}] failed to delete files in [{}, {}): {:?}",
+                    region_id,
+                    escape(&start_key),
+                    escape(&end_key),
+                    e
+                );
+                return;
+            }
+        }
         if let Err(e) =
             util::delete_all_in_range(&self.kv_db, &start_key, &end_key, self.use_delete_range)
         {
             error!(
-                "failed to delete data in [{}, {}): {:?}",
+                "[region {}] failed to delete data in [{}, {}): {:?}",
+                region_id,
                 escape(&start_key),
                 escape(&end_key),
                 e
             );
+        } else {
+            info!(
+                "[region {}] succeed in deleting data in [{}, {})",
+                region_id,
+                escape(&start_key),
+                escape(&end_key),
+            );
+        }
+    }
+
+    fn cleanup_overlap_ranges(&mut self, start_key: &[u8], end_key: &[u8]) {
+        let overlap_ranges = self.pending_delete_ranges
+            .drain_overlap_ranges(start_key, end_key);
+        for (region_id, s_key, e_key) in overlap_ranges {
+            self.cleanup_range(region_id, s_key, e_key, false /* use_delete_files */);
+        }
+    }
+
+    fn insert_pending_delete_range(
+        &mut self,
+        region_id: u64,
+        start_key: Vec<u8>,
+        end_key: Vec<u8>,
+    ) -> bool {
+        if self.clean_stale_peer_delay.as_secs() == 0 {
+            return false;
+        }
+
+        self.cleanup_overlap_ranges(&start_key, &end_key);
+
+        info!(
+            "[region {}] register deleting data in [{}, {})",
+            region_id,
+            escape(&start_key),
+            escape(&end_key),
+        );
+        let timeout = time::Instant::now() + self.clean_stale_peer_delay;
+        self.pending_delete_ranges
+            .insert(region_id, start_key, end_key, timeout);
+        true
+    }
+
+    fn clean_timeout_ranges(&mut self) {
+        STALE_PEER_PENDING_DELETE_RANGE_GAUGE.set(self.pending_delete_ranges.len() as f64);
+
+        let now = time::Instant::now();
+        let mut timeout_ranges = self.pending_delete_ranges.drain_timeout_ranges(now);
+        for (region_id, start_key, end_key) in timeout_ranges.drain(..) {
```

Comment:
What if we apply a snapshot after we have drained the timeout ranges but before we have cleaned up the ranges?